src/cls/rbd/cls_rbd.cc

```diff
@@ -6898,18 +6900,135 @@ int dir_remove(cls_method_context_t hctx,
 }
 
 static const string RBD_GROUP_SNAP_KEY_PREFIX = "snapshot_";
+static const string RBD_GROUP_SNAP_ORDER_KEY_PREFIX = "snap_order_";
+static const string RBD_GROUP_SNAP_MAX_ORDER_KEY = "snap_max_order";
 
 std::string snap_key(const std::string &snap_id) {
   ostringstream oss;
   oss << RBD_GROUP_SNAP_KEY_PREFIX << snap_id;
   return oss.str();
 }
 
-int snap_list(cls_method_context_t hctx, cls::rbd::GroupSnapshot start_after,
-              uint64_t max_return,
-              std::vector<cls::rbd::GroupSnapshot> *group_snaps)
+std::string snap_order_key(const std::string &snap_id) {
+  ostringstream oss;
+  oss << RBD_GROUP_SNAP_ORDER_KEY_PREFIX << snap_id;
+  return oss.str();
+}
+
+static const string snap_id_from_order_key(const string &key) {
+  return key.substr(RBD_GROUP_SNAP_ORDER_KEY_PREFIX.size());
+}
+
+int snap_list_sorted(cls_method_context_t hctx, cls::rbd::GroupSnapshot start_after,
+                     uint64_t max_return,
+                     std::vector<cls::rbd::GroupSnapshot> *group_snaps)
 {
   int max_read = RBD_MAX_KEYS_READ;
+  std::map<std::string, bufferlist> vals;
+  std::string last_read = snap_key("");
```

Comment:
I'm not sure I see the benefit here - the client would still need to do the same operations as the osd and would still need to read the entire list every time as it cannot cache a previously read value in order to provide batch support.
We would still need to read all the order keys and return them as the order info is the value. The map will not preserve ordering.

Reply:
> I'm not sure I see the benefit here - the client would still need to do the same operations as the osd

Right.

> and would still need to read the entire list every time as it cannot cache a previously read value in order to provide batch support.

No -- if the client is issuing reads in batches, the OSD would read exactly that batch.  This is the key difference to the current code.

> We would still need to read all the order keys and return them as the order info is the value.

Right, but both the snapshot keys and the order keys would be read just once.

Consider a scenario where there are 3 snapshots with order keys and the client issuing reads with `max_return=1`.  The on-disk state is as follows:

`snapshot_id1`      (name "snap3", created third)
`snapshot_id2`      (name "snap1", created first)
`snapshot_id3`      (name "snap2", created second)
`snap_order_id1`    3
`snap_order_id2`    1
`snap_order_id3`    2

With the current code:

- client calls `snap_list_sorted` with `start_after=""` and `max_return=1`
- OSD reads all 6 keys, decodes everything, performs the sort, erases snap2 and snap3 from the vector and returns snap1
- client calls `snap_list_sorted` with `start_after="id2"` and `max_return=1`
- OSD reads all 6 keys, decodes everything, performs the sort, erases snap1 and snap3 from the vector and returns snap2
- client calls `snap_list_sorted` with `start_after="id3"` and `max_return=1`
- OSD reads all 6 keys, decodes everything, performs the sort, erases snap1 and snap2 from the vector and returns snap3

With my proposal (`snap_list_order` is the provisionary name of the new method):

- client calls `snap_list` with `start_after=""` and `max_return=1`
- OSD reads 1 key, decodes it and returns snap3
- client calls `snap_list` with `start_after="id1"` and `max_return=1`
- OSD reads 1 key, decodes it and returns snap1
- client calls `snap_list` with `start_after="id2"` and `max_return=1`
- OSD reads 1 key, decodes it and returns snap2
- client calls `snap_list_order` with `start_after=""` and `max_return=1`
- OSD reads 1 key, decodes it and returns `"id1": 3` pair
- client calls `snap_list_order` with `start_after="id1"` and `max_return=1`
- OSD reads 1 key, decodes it and returns `"id2": 1` pair
- client calls `snap_list_order` with `start_after="id2"` and `max_return=1`
- OSD reads 1 key, decodes it and returns `"id3": 2` pair
- at this point the client has all snaps and all order pairs and can perform the sort

Note how at each step we read and decoded exactly what the client told us to and performed the sort just once.  We doubled the number of round-trips to the OSD, but that is a fixed, linear, overhead.  With the current code the overhead is not fixed and quadratic instead of linear.

Reply:
> We doubled the number of round-trips to the OSD

Also note that:

- This is needed only if sorting is required -- otherwise `snap_list_order` won't be called at all.
- This is essentially the price for keeping `GroupSnapshot` struct intact (and the associated simplicity).  If we wanted to go that route, a second set of round-trips to the OSD wouldn't be needed even for sorting.

Reply:
I am not in favour of having the client do any of this processing for the reasons in an earlier comment. We do not have any control over the clients in case we need to change something.  I think it is better to keep the processing on the OSD. The above algo also implies that the storage format is not completely opaque to the client if it knows it has to get the sorted order separately.

Reply:
> * This is essentially the price for keeping `GroupSnapshot` struct intact (and the associated simplicity).

We could avoid doubling the number of round-trips to the OSD while still not touching `GroupSnapshot` struct by introducing method that would return both snapshot information _and_ order information.  For example, a map from group snap ID to a pair of `GropSnapshot` and order.  This would probably be the most efficient of the "easy" approaches.

Reply:
> I am not in favour of having the client do any of this processing for the reasons in an earlier comment. We do not have any control over the clients in case we need to change something.

Depending on how that "something" changes the semantics of the method, client-side changes may be needed anyway.

> I think it is better to keep the processing on the OSD.

I disagree.  Even if we don't expect tons of group snapshots, adding a method which pretends to be efficient but in reality has quadratic performance characteristics is wrong.

> The above algo also implies that the storage format is not completely opaque to the client if it knows it has to get the sorted order separately.

I don't think it would become any less opaque than it is today (it's the client who is in charge of `start_after` and `max_return` parameters, after all).  From the client's perspective, it's just a (new) method that they can call to get the order/sequence number of the snapshot.  The client doesn't get exposed to how the sequence number is stored on-disk.

Reply:
> > I am not in favour of having the client do any of this processing for the reasons in an earlier comment. We do not have any control over the clients in case we need to change something.
> 
> Depending on how that "something" changes the semantics of the method, client-side changes may be needed anyway.

Client side changes in a well designed call should be limited to changes in the output format like additional fields etc - it should not include actually doing the processing like sorting.
> 
> > I think it is better to keep the processing on the OSD.
> 
> I disagree. Even if we don't expect tons of group snapshots, adding a method which pretends to be efficient but in reality has quadratic performance characteristics is wrong.

It is not really quadratic in a real world scenario - the client currently sends a max_read value of 1024 so it is extremely likely that there will be very few calls. The algorithm as such is O(n +nlogn).

> 
> > The above algo also implies that the storage format is not completely opaque to the client if it knows it has to get the sorted order separately.
> 
> I don't think it would become any less opaque than it is today (it's the client who is in charge of `start_after` and `max_return` parameters, after all). From the client's perspective, it's just a (new) method that they can call to get the order/sequence number of the snapshot. The client doesn't get exposed to how the sequence number is stored on-disk.

The client is requesting a sorted list as such should be able to get one back from the OSD without having to do it on its own. start_after and max_return are valid input in terms of listing - it does not imply it knows how it is stored.

Reply:
> Client side changes in a well designed call should be limited to changes in the output format like additional fields etc - it should not include actually doing the processing like sorting.

I disagree with this statement, both in the context of this PR and in general.  The same collection can be sorted in multiple different ways, be partially sorted (think top-k, etc) or not be sorted at all.  Having the server spend time and resources sorting the entire collection in one particular way can be very suboptimal.

> It is not really quadratic in a real world scenario - the client currently sends a max_read value of 1024 so it is extremely likely that there will be very few calls. The algorithm as such is O(n +nlogn).

We are adding a CLS method -- a thing that, unlike some API internal to librbd, is very hard to change/improve after it lands.  Even if "very few" is going to be limited to 2-3, there is no excuse for a design that would perform the _same_ read/decode/sort dance 2-3 times instead of just once per listing (and, to top it off, do that on the OSD where CPU cycles are most precious!).

> The client is requesting a sorted list as such should be able to get one back from the OSD without having to do it on its own.

Already disagreed with above ;)

> start_after and max_return are valid input in terms of listing - it does not imply it knows how it is stored.

It kind of does -- the reason that these parameters are exposed to the client is that the server is able to _efficiently_ pick up where it left off.  Consequently, the client can assume that the on-disk format allows for that and that there is some sort of iterator in place.  Server-side sorting basically invalidates that iterator and compensating for that by erasing "unneeded" entries from the vector just before returning is bad.